### PR TITLE
fix(inbox): durable response numbering — counter out of the sync mirror + four floors

### PR DESCRIPTION
**Pull request by WingedGuardian**
## Problem

The inbox response writer numbers its output files monotonically (`{stem}-{N}.genesis.md`) with a "persistent counter file as primary source of truth… numbers never go backwards" guarantee. That counter file lived **inside the vault-sync mirror directory**, where the 5-minute sync deleted it **every cycle** (measured 175 deletions in the sync log) — the guarantee silently rested on the filesystem scan alone. When a sync false-positive wiped every response file in the watch dir (2026-07-29 incident, fixed separately in #1448), numbering restarted at 1 and weeks of subsequent responses were delivered as **silent overwrites** of already-read files in the user's vault.

## What this does

- **Counter store relocated** out of the mirror to `~/.genesis/state/inbox-counters.json` (`{directory: {stem: high_water}}`) — nothing else may delete it.
- **Four max-wins floors**: durable store → legacy in-mirror file (migration for installs upgrading mid-sequence) → watch-dir scan → **new DB floor** over `inbox_items.response_path` history (read-only connection; degrades to 0 with a warning — a response write never fails on a DB hiccup). Even total loss of the first three floors can't renumber backwards.
- **Off-loop + locked**: counter derivation moves off the event loop (`asyncio.to_thread`) and is serialized on a module lock — off-loop execution forfeits coroutine atomicity, and the unlocked read-modify-write could mint duplicate numbers. Belt: `write_response` raises loudly if its freshly-minted target already exists rather than letting `os.replace` clobber a delivered response.
- **Corruption blast-radius**: a corrupt individual store value zeroes only that stem's floor; only an unreadable/undecodable file resets the parsed store — one bad value can no longer destroy every other directory's high-water mark on the next persist.
- **Test hygiene**: new `tests/test_inbox/conftest.py` autouse fixture isolates the counter store for the whole inbox suite (same durable-state class as the top-level alert-queue isolation).

No interface change: `write_response` signature/return identical; the single runtime call site is unaffected.

## Testing

20 writer tests (9 new: watch-dir wipe survival, DB-floor rescue/strictness, unreadable-DB degradation, legacy migration, concurrency, corrupt-value blast radius, flat-legacy-copy safety) — the concurrency and corrupt-value tests verify-RED by in-place revert; 194 writer-consumer tests (monitor/drop dispatch/edge cases/scanner) green; ruff clean.

## Review

Pre-push genesis-architect adversarial pass: 2 SHOULD-FIX found and fixed (the `to_thread` atomicity trade needing the module lock; the corrupt-value store wipe), each locked by a RED-verified test. DB-floor connect/create semantics verified empirically by the reviewer.

Ledger: 852664d365284036bf75d9406ba7b8e9

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
**Automated first-pass review** (structural patterns + PII/secrets/wording).

Note: this PR is authored by the same account this review runs as, so GitHub rejects a formal `REQUEST_CHANGES` review here (self-review restriction) — posting as a comment instead. The finding below is still a structural ERROR and should be addressed before merge.

## Structural review

**ERROR — Raw SQL against genesis.db, bypassing `genesis.db.crud`** (`src/genesis/inbox/writer.py:135-166`, new `_db_floor()`)

```python
db_path = env.genesis_db_path()
conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
...
rows = conn.execute(
    "SELECT response_path FROM inbox_items WHERE response_path IS NOT NULL",
).fetchall()
```

Two established alternatives already exist for exactly this query:
- `src/genesis/db/crud/inbox_items.py` — the project's crud module for this table, with `get_all_known`, `get_recent_completed`, etc. already selecting `response_path`.
- `src/genesis/db/connection.py:open_ro_connection()` — the project's existing standalone read-only connection helper (`mode=ro`, WAL-aware — same rationale documented there as the inline comment in `_db_floor`), instead of hand-rolling `sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)` again here.

This raw synchronous `sqlite3` query direct against `inbox_items` is the "Raw SQL against genesis.db" pattern the project's conventions call out (crud modules only, except in migrations). I recognize `_next_counter`/`_db_floor` run inside `asyncio.to_thread` (writer.py:104), which is presumably why a sync `sqlite3` connection was reached for instead of the async `aiosqlite`-based crud/`open_ro_connection` path — but that constraint isn't noted anywhere in the PR description or code comments, and as written it reintroduces a second, divergent way to open a read-only connection to `genesis.db`. Please either route this through the crud module (adding a small sync/thread-safe accessor there if the async signature is genuinely incompatible with the `to_thread` call site) or document why the existing helpers can't be reused here.

Everything else in the diff looks clean against the structural checklist: no `nohup`, no direct `task_states` writes, no `/tmp/` usage, all `except` clauses are typed, no mutable default args, imports are stdlib-grouped correctly, new private helpers are documented, and the new/changed behavior is covered by an extensive new test suite (`tests/test_inbox/test_writer.py`, `tests/test_inbox/conftest.py`).

## PII / Secrets / Wording scan

CLEAN — no secrets, credentials, private IPs, or hardcoded personal paths in the diff. PR title/body describe the incident (vault-sync deletion of a counter file, silent response overwrites) without naming private infrastructure, credentials, or exploitable specifics — no wording concerns.

---


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
**Structural review — first pass** (patterns/conventions only; not a logic review)

Scope: `src/genesis/inbox/writer.py`, `tests/test_inbox/conftest.py` (new), `tests/test_inbox/test_writer.py`.

No ERROR-level pattern violations found (no bare `except:`, no mutable defaults, no `/tmp/` usage, no `nohup`, no direct `task_states` writes, no `git add -A`/force-push patterns, no `os.killpg`).

**WARNING**

1. **Raw SQL against `genesis.db` outside `db/crud`** — `src/genesis/inbox/writer.py:154-159` (`_db_floor`) opens a direct `sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)` and hand-writes `SELECT response_path FROM inbox_items ...` instead of going through `genesis.db.crud.inbox_items`. There's codebase precedent for this exact read-only-connection idiom outside `crud/` (`src/genesis/memory/health.py:194`), and the sync/thread context (`_db_floor` runs inside `asyncio.to_thread`) makes the async `aiosqlite`-based crud layer awkward to reuse directly — so this isn't blocking. But for consistency/testability it'd be worth either (a) adding a small sync-safe accessor in `db/crud/inbox_items.py` this can call, or (b) a one-line comment noting the precedent so a future reviewer doesn't have to rediscover it.

2. **Duplicated local imports of `genesis.env`** — `writer.py:46` (`_counter_store_path`) and `writer.py:149` (`_db_floor`) both do `from genesis import env` inside the function body, even though the module already imports `from genesis.env import user_timezone` at the top (`writer.py:26`) — i.e., importing `genesis.env` at module scope is already proven safe here (no circular-import issue). Minor, but these two locals could just use the existing top-level import (or import `genesis_home`/`genesis_db_path` alongside `user_timezone`).

3. **New `RuntimeError` invariant-breach path is untested** — `writer.py:106-113` adds a new branch: if the freshly-minted `target` already exists, `write_response` now raises `RuntimeError` instead of silently overwriting. The PR's test suite is otherwise very thorough (9 new tests covering DB-floor rescue, corruption blast-radius, concurrency, legacy migration, etc.), but none of them exercise this specific raise (e.g. by pre-creating the target file and asserting `pytest.raises(RuntimeError)`). Worth a test given it's the "belt" half of the belt-and-suspenders fix described in the PR body.

**NOTE**

- Import grouping/ordering, docstrings, and commit message (`fix(inbox): ...`) all look correct — no findings there.

Everything else (four-floor precedence logic, lock placement, store corruption handling, migration path) is well-covered by the new tests and out of scope for a structural pass.




---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/genesis/inbox/writer.py [hunk 1]
```diff
@@ -99,25 +132,111 @@ def _write_atomic(tmp_path: Path, target: Path, content: str) -> None:
...
+      4. the DB's ``inbox_items.response_path`` history (survives even a
+         total loss of 1-3).
+
+    Runs off the event loop (write_response wraps it in to_thread), so the
+    store read-modify-write is serialized on a module lock — without it two
+    concurrent calls could mint the same number and silently overwrite.
     """
-    counter_path = directory / _COUNTER_FILE
+    with _COUNTER_LOCK:
+        return _next_counter_locked(directory, base_name, suffix)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Fixed in 2f3b97fb. The module lock only serialized threads in one interpreter; allocation is now an atomic cross-process reservation via os.link (create-or-fail; a second writer process can't mint a duplicate), verified with 8 real subprocesses incl. the store bookkeeping permanently broken. An independent by-execution data-integrity pass confirmed no reuse across processes.

File: src/genesis/inbox/writer.py [hunk 2]
```diff
@@ -99,25 +132,111 @@ def _write_atomic(tmp_path: Path, target: Path, content: str) -> None:
...
+
+    prefix = f"{base_name}-"
+    floor = 0
+    try:
+        db_path = env.genesis_db_path()
+        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
+        try:
+            rows = conn.execute(
+                "SELECT response_path FROM inbox_items WHERE response_path IS NOT NULL",
+            ).fetchall()
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py:218]: Addressed in 2f3b97fb, though the resolution flipped: the DB floor is now consulted UNCONDITIONALLY (the earlier skip-gate under-counted on a stale-but-nonzero partial restore, re-minting a delivered number). The O(history) scan is consciously accepted — bounded, off the event loop, sub-ms at personal inbox scale — because correctness outweighs the micro-optimization on a silent-data-loss path.

File: src/genesis/inbox/writer.py [hunk 3]
```diff
@@ -99,25 +132,111 @@ def _write_atomic(tmp_path: Path, target: Path, content: str) -> None:
...
+    # skips numbers (cosmetic gaps) and inherently covers directory renames;
+    # under-counting is the overwrite bug. Kept for signature symmetry.
+    del directory
+    from genesis import env
+
+    prefix = f"{base_name}-"
+    floor = 0
+    try:
+        db_path = env.genesis_db_path()
+        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Fixed in an earlier commit on this branch: the read-only URI is now built with urllib.parse.quote(str(db_path), safe='/'), so a path containing ? / # / % is percent-encoded instead of parsed as query/fragment. Locked by test_db_floor_reads_path_with_uri_reserved_char (DB at an 'a#b/' path).

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/inbox/writer.py [hunk 4]
```diff
@@ -75,49 +103,186 @@ async def write_response(
...
+        )
+        tmp_path = Path(tmp_name)
+        try:
+            with os.fdopen(fd, "w", encoding="utf-8") as fh:
+                fh.write(body)
+            for _ in range(_ALLOC_MAX_RETRIES):
+                next_num = _next_counter(base_dir, stem, suffix)
+                target = base_dir / f"{stem}-{next_num}{suffix}"
+                try:
+                    os.link(str(tmp_path), str(target))  # atomic; ENOENT-safe
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Resolved two ways in 766aac88: (1) the branch now merges current main, so inbox_sync.sh carries #1448's `--exclude ".genesis-*"` — the .genesis-tmp-* allocation temp matches that glob and is protected from the sync; (2) the flock is removed entirely, so the window between mkstemp and os.link no longer spans a lock acquire + full DB scan under a held lock.

File: src/genesis/inbox/writer.py [hunk 5]
```diff
@@ -75,49 +103,186 @@ async def write_response(
...
+         total loss of 1-3).
+
+    Runs off the event loop (write_response wraps it in to_thread), so the
+    store read-modify-write is serialized on a module lock — without it two
+    concurrent calls could mint the same number and silently overwrite.
     """
-    counter_path = directory / _COUNTER_FILE
+    store_path = _counter_store_path()
+    store_path.parent.mkdir(parents=True, exist_ok=True)
+    lock_path = store_path.with_name(store_path.name + ".lock")
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Fixed in 766aac88 — you're right, I'd made the store load-bearing. Root fix: the counter store + lock are now best-effort only. The fcntl flock is removed (redundant: the DB delivery log is the authority and os.link is the cross-process guard), and the store mkdir/write are inside the persist try/except, so a read-only ~/.genesis/state degrades to the DB/disk floors + os.link and the response is still written. Verify-RED: the old version RAISES PermissionError on a read-only store dir; the new one writes doc-1. Locked by test_write_survives_unwritable_counter_store.

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/inbox/writer.py [hunk 6]
```diff
@@ -75,49 +102,177 @@ async def write_response(
...
+            conn.close()
+    except (sqlite3.Error, OSError) as exc:
+        logger.warning(
+            "Counter DB floor unavailable (%s): falling back to store/scan floors",
+            exc,
+            exc_info=True,
+        )
+        return 0
+    for (rp,) in rows:
+        name = Path(rp).name
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Fixed: _db_floor wraps the per-row parse (except TypeError/ValueError/OSError), so a BLOB/non-str response_path is skipped, never crashing the floor. Locked by test_db_floor_skips_malformed_row.

File: src/genesis/inbox/writer.py [hunk 7]
```diff
@@ -75,49 +102,177 @@ async def write_response(
...
+        base_dir: Path, stem: str, suffix: str, body: str,
+    ) -> Path:
+        # Write the content to a unique tmp in the SAME dir (same fs → link is
+        # atomic and the target appears fully-formed, never partial). The
+        # ".genesis-" prefix means a crash-leaked tmp is BOTH scanner-ignored
+        # (dot-prefixed) AND excluded from the vault sync (--exclude ".genesis-*")
+        # — it can never leak into the user's Obsidian folder.
+        fd, tmp_name = tempfile.mkstemp(
+            dir=base_dir, prefix=".genesis-tmp-", suffix=".tmp",
+        )
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py:145]: Fixed + hardened by a whole-class audit: response now carries umask perms, but the umask is snapshotted once at import (os.umask is process-global and raced per-write on the to_thread pool) and the chmod is best-effort (a chmod-failing fs must not fail the write). Locked by test_response_file_has_standard_perms + test_concurrent_writes_do_not_touch_process_umask + test_chmod_failure_does_not_fail_write.

File: src/genesis/inbox/writer.py [hunk 8]
```diff
@@ -75,49 +102,177 @@ async def write_response(
...
+                    os.link(str(tmp_path), str(target))  # atomic; ENOENT-safe
+                except FileExistsError:
+                    continue  # another writer took this number — re-derive
+                return target
+            raise RuntimeError(
+                f"could not allocate a unique response number for {stem!r} "
+                f"after {_ALLOC_MAX_RETRIES} attempts",
+            )
+        finally:
+            tmp_path.unlink(missing_ok=True)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py]: Fixed: the tmp unlink in the finally is best-effort now, so a cleanup failure can never mask an already-published response. Locked by test_tmp_cleanup_failure_does_not_mask_publish.

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/inbox/writer.py [hunk 9]
```diff
@@ -75,49 +110,199 @@ async def write_response(
...
+    if store_path.exists():
+        try:
+            loaded = json.loads(store_path.read_text(encoding="utf-8"))
+            store = loaded if isinstance(loaded, dict) else {}
+        except (json.JSONDecodeError, OSError):
+            store = {}  # Unreadable/corrupt file — fall through to other floors
+        try:
+            stored_max = int(store.get(dir_key, {}).get(base_name, 0))
+        except (ValueError, TypeError, AttributeError):
+            stored_max = 0  # Corrupt value — keep the rest of the store intact
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py:293]: Consciously accepted (documented, no change): single-uid lxc container, same-user vault. Python ints don't overflow (and the store parse is already guarded), and no POSIX default ACLs are configured on ~/inbox. Not worth changing already-reviewed code at the escalation cap; the correctness core (DB floor + os.link) is unaffected.

File: src/genesis/inbox/writer.py [hunk 10]
```diff
@@ -75,49 +110,199 @@ async def write_response(
...
+        try:
+            with os.fdopen(fd, "w", encoding="utf-8") as fh:
+                fh.write(body)
+            # mkstemp forces 0o600; the published response (a hardlink to this
+            # inode) should carry the perms an ordinary write would (umask-
+            # derived, from the import-time snapshot — never read os.umask here,
+            # it is process-global and races on the to_thread pool). Best-effort:
+            # a chmod failure (e.g. a network/FUSE fs) must NEVER fail the write.
+            try:
+                os.chmod(tmp_path, 0o666 & ~_UMASK)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/inbox/writer.py:156]: Consciously accepted (documented, no change): single-uid lxc container, same-user vault. Python ints don't overflow (and the store parse is already guarded), and no POSIX default ACLs are configured on ~/inbox. Not worth changing already-reviewed code at the escalation cap; the correctness core (DB floor + os.link) is unaffected.
